internals/overlord/checkstate/handlers.go

```diff
@@ -0,0 +1,207 @@
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package checkstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+func (m *CheckManager) doPerformCheck(task *state.Task, tomb *tomb.Tomb) error {
+	var details checkDetails
+	m.state.Lock()
+	changeID := task.Change().ID()
+	err := task.Get("check-details", &details)
+	m.state.Unlock()
+	if err != nil {
+		return fmt.Errorf("cannot get check details for task %q: %v", task.ID(), err)
+	}
+
+	m.planLock.Lock()
+	plan := m.plan
+	m.planLock.Unlock()
+	config, ok := plan.Checks[details.Name]
+	if !ok {
+		logger.Debugf("Check %q no longer exists in plan.", details.Name)
+		return nil
+	}
+
+	m.checksLock.Lock()
+	data := m.checks[config.Name]
+	if data == nil {
+		data = &checkData{}
+		m.checks[config.Name] = data
+	}
+	data.config = config
+	data.cancel = func() { tomb.Kill(nil) }
+	data.changeID = changeID
+	m.checksLock.Unlock()
+
+	logger.Debugf("Performing check %q with period %v", details.Name, config.Period.Value)
+	ticker := time.NewTicker(config.Period.Value)
+	defer ticker.Stop()
+
+	chk := newChecker(config, plan)
+	for {
+		select {
+		case <-ticker.C:
+			err := runCheck(tomb.Context(nil), config, chk, config.Timeout.Value)
+			if err != nil {
+				// Check failed, switch to recovering a failing check.
+				failures, actionRan := m.handleCheckFailure(config, err)
+
+				// Hold checksLock while updating state so that if someone
+				// calls Checks() they always see the latest state. We must
+				// always acquire checksLock and then state lock in that order.
+
+				m.checksLock.Lock()
+
+				m.state.Lock()
+				newChangeID := m.recoverCheckChange(config)
+				m.state.EnsureBefore(0)
+				m.state.Unlock()
+
+				data.failures = failures
+				data.actionRan = actionRan
+				data.changeID = newChangeID
+				m.checksLock.Unlock()
+
+				// Returning the error means perform-check goes to Error status
+				// and logs the error to the task log.
+				return err
+			}
+
+		case <-tomb.Dying():
+			logger.Debugf("Check %q stopped during perform-check: %v", config.Name, tomb.Err())
+			return tomb.Err()
+		}
+	}
+}
+
+func runCheck(ctx context.Context, config *plan.Check, chk checker, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
```

Comment:
For my understanding: I would have assumed that the ```check(ctx)``` is blocking and only returns when it is done. The context will be cancelled when the tomb dies, but what is the purpose of having using the ```cancel``` on function exit after the check completed or failed ?

Reply:
Yep, `check(ctx)` is blocking and only returns when it is done. The `WithTimeout` context will be cancelled when the parent (tomb) dies or when the timeout occurs. The `cancel()` call is to free resources, per the [docs](https://pkg.go.dev/context#WithDeadline):

> Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this [Context](https://pkg.go.dev/context#Context) complete.

Also from the example:

```
// Even though ctx will be expired, it is good practice to call its
// cancellation function in any case. Failure to do so may keep the
// context and its parent alive longer than necessary.
defer cancel()
```